[NFC] SIL: Clarified deinit barrier APIs.

nate-chandler opened the pull request:
This PR adds three commits on top of https://github.com/apple/swift/pull/70773 .

An instruction is a deinit barrier whenever one of three component predicates is true for it.  In the case of applies, it is true whenever one of those three predicates is true for any of the instructions in any of its callees; that fact is cached in the side-effect analysis of every function.

If side-effect analysis or callee analysis is unavailable, in order to define each of those three component predicates on a `FullApplySite`/`EndApplyInst`/`AbortApplyInst`, it would be necessary to define them to conservatively return true: it isn't known whether any of the instructions in any of the callees were deinit barriers.

Refactored the two versions of the deinit barrier predicate (namely `Instruction.isDeinitBarrier(_:)` and `swift::mayBeDeinitBarrierNotConsideringSideEffects`) to handle `FullApplySite`/`EndApplyInst`/`AbortApplyInst`s specially first (to look up the callees' side-effect and to conservatively bail, respectively).  Asserted that the three component predicates are not called with `FullApplySite`/`EndApplyInst`/`AbortApplyInst`s.  Callers should instead use the `isDeinitBarrier` APIs.

An alternative would be to conservatively return true from the three components.  That seems more likely to result in direct calls to these member predicates, however, and at the moment at least there is no reason for such calls to exist.  If some other caller besides the deinit-barrier predicates needs to call this function, side-effect analysis should be updated to cache these three properties separately at that point.

nate-chandler committed:
[SIL] Builtins that read/write mem may access ptr.

When an instruction `mayReadOrWriteMemory`, the `mayAccessPointer`
predicate relies on the `visitAccessedAddress` utility to determine
whether an instruction may access a pointer.  That utility doesn't (and
can't, without changing the representation of the builtin) visit
`RawPointer` operands as used by memmove/memcpy builtins.

Previously, this resulted `mayAccessPointer` returning `false` for such
builtins.  Here, this is fixed by making the predicate handle
`BuiltinInsts` separately.  Instead of just always returning `true` in
the face of a builtin, rely on the BuiltinInfo associated with a
BuiltinInst and use `mayReadOrWriteMemory`.

rdar://120656227

nate-chandler committed:
[NFC] SIL: Renamed maySynchronize.

Dropped the NotConsideringSideEffects suffix.

nate-chandler committed:
[NFC] SIL: Clarified deinit barrier APIs.

An instruction is a deinit barrier whenever one of three component
predicates is true for it.  In the case of applies, it is true whenever
one of those three predicates is true for any of the instructions in any
of its callees; that fact is cached in the side-effect analysis of every
function.

If side-effect analysis or callee analysis is unavailable, in order to
define each of those three component predicates on a
`FullApplySite`/`EndApplyInst`/`AbortApplyInst`, it would be necessary
to define them to conservatively return true: it isn't known whether any
of the instructions in any of the callees were deinit barriers.

Refactored the two versions of the deinit barrier predicate (namely
`Instruction.isDeinitBarrier(_:) and
`swift::mayBeDeinitBarrierNotConsideringSideEffects`) to handle
`FullApplySite`/`EndApplyInst`/`AbortApplyInst`s specially first (to
look up the callees' side-effect and to conservatively bail,
respectively).  Asserted that the three component predicates are not
called with `FullApplySite`/`EndApplyInst`/`AbortApplyInst`s.  Callers
should instead use the `isDeinitBarrier` APIs.

An alternative would be to conservatively return true from the three
components.  That seems more likely to result in direct calls to these
member predicates, however, and at the moment at least there is no
reason for such calls to exist.  If some other caller besides the
deinit-barrier predicates needs to call this function, side-effect
analysis should be updated to cache these three properties separately at
that point.

atrick reviewed (approved):
LGTM; although in all three APIs, I think Builtins that mayReadOrWriteMemory should be treated conservatively.

nate-chandler committed:
[NFC] SIL: Define deinit barrier preds on builtins

The three predicates are all now defined and defined correctly on
builtins to return true whenever mayReadOrWriteMemory is true.

nate-chandler commented:
@swift-ci please test
